### PR TITLE
[draft] Bulk update 'waiting' tasks instead of individual save

### PR DESCRIPTION
**Pull request by kdelee**
Previously, was seeing able to schedule ~ 20 jobs/second

After, able to schedule ~ 150 jobs/second

We already have precedent for skipping `task.save()` for performance reasons (see previous code that was doing this for job_explanation).
Now we can just do it for all jobs going into `waiting`

##### ISSUE TYPE

 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME

 - API

##### ADDITIONAL INFORMATION

Sometimes you want to schedule MORE jobs FASTER

### CODE REVIEW

File: awx/main/scheduler/task_manager.py [hunk 1]
```diff
@@ -446,7 +446,8 @@ def process_job_dep_failures(self, task):
                     dep.name,
                     dep.id,
                 )
-                task.save(update_fields=['status', 'job_explanation'])
+                # save now occurs in process_pending tasks at end of loop
+                # task.save(update_fields=['status', 'job_explanation'])
```
Comment (kdelee) [inline awx/main/scheduler/task_manager.py]: maybe we need this one to deal with updating parents :expressionless:

**Comment by AlanCoding** [comment]
I enabled the Django debug toolbar, and put this canary script into the root view, and then loaded the page and isolated the queries from the save.

```python
from awx.main.models import Job
job = Job.objects.first()
job.save()
```

This results in _18 queries_ to finish the update.

I repeated the same process, with an adjusted script:

```python
from awx.main.models import Job
job.status = "running"
job = Job.objects.first()
job.save()
```

In this case, the prior status was "successful". This results in _42 queries_ to do the save.

The 7.5x improvement you found isn't so surprising with these considered. We talked about https://github.com/ansible/awx/pull/13553, and yes that is a part of the noise seen, but when we go to the update to the running status, the queries look like exactly what was intended to be addressed in https://github.com/ansible/awx/pull/13124. The performance killer there is:

 - job status goes to running
   - ok, need to update current job of JT
     - oh, JT `organization` field might be changing, double-check all the role parents

Even though we have the "new" role system in place, the code paths for this is still in place. Looking at the diff from 13124, that is way more easily achievable, where the criteria for RBAC-sensitive changes is simply "is the parent field in update_fields?"

**Comment by kdelee** [comment]
FWIW api tests failing now -- probably not updating parent when deleting in `waiting` ? need to investigate

**Comment by kdelee** [comment]
> Even though we have the "new" role system in place, the code paths for this is still in place. Looking at the diff from 13124, that is way more easily achievable, where the criteria for RBAC-sensitive changes is simply "is the parent field in update_fields?"

@AlanCoding so are you saying that now with move to django-ansible-base, you'd like to ressurect #13124 ?

**Comment by AlanCoding** [comment]
The https://github.com/ansible/awx/pull/13124 patch modifies code that is intended to update role parents IIRC. I have it concretely planned to delete the role ancestors relationship. I have not thought very much about whether we could delete the parents relationship. If we delete the relationship, then we wouldn't optimize the code involved there, we would delete it. There is a whole bunch on the agenda that has gotten delayed due to deadline pressures.

**Comment by kdelee** [comment]
running same workload as https://github.com/ansible/awx/pull/15174, getting around ~ 142 jobs/second (_schedule + commit seconds)

```
# HELP task_manager_get_tasks_seconds Time spent in loading tasks from db
# TYPE task_manager_get_tasks_seconds gauge
task_manager_get_tasks_seconds{node="awx-1"} 17.69417297898326
# HELP task_manager_start_task_seconds Time spent starting task
# TYPE task_manager_start_task_seconds gauge
task_manager_start_task_seconds{node="awx-1"} 36.96652986994013
# HELP task_manager_process_running_tasks_seconds Time spent processing running tasks
# TYPE task_manager_process_running_tasks_seconds gauge
task_manager_process_running_tasks_seconds{node="awx-1"} 1.6810372471809387e-06
# HELP task_manager_process_pending_tasks_seconds Time spent processing pending tasks
# TYPE task_manager_process_pending_tasks_seconds gauge
task_manager_process_pending_tasks_seconds{node="awx-1"} 39.52465379599016
# HELP task_manager__schedule_seconds Time spent in running the entire _schedule
# TYPE task_manager__schedule_seconds gauge
task_manager__schedule_seconds{node="awx-1"} 70.74675493501127
# HELP task_manager__schedule_calls Number of calls to _schedule, after lock is acquired
# TYPE task_manager__schedule_calls gauge
task_manager__schedule_calls{node="awx-1"} 5
# HELP task_manager_recorded_timestamp Unix timestamp when metrics were last recorded
# TYPE task_manager_recorded_timestamp gauge
task_manager_recorded_timestamp{node="awx-1"} 0.0
# HELP task_manager_tasks_started Number of tasks started
# TYPE task_manager_tasks_started gauge
task_manager_tasks_started{node="awx-1"} 10000
# HELP task_manager_running_processed Number of running tasks processed
# TYPE task_manager_running_processed gauge
task_manager_running_processed{node="awx-1"} 0
# HELP task_manager_pending_processed Number of pending tasks processed
# TYPE task_manager_pending_processed gauge
task_manager_pending_processed{node="awx-1"} 39197
# HELP task_manager_tasks_blocked Number of tasks blocked from running
# TYPE task_manager_tasks_blocked gauge
task_manager_tasks_blocked{node="awx-1"} 67
# HELP task_manager_commit_seconds Time spent in db transaction, including on_commit calls
# TYPE task_manager_commit_seconds gauge
task_manager_commit_seconds{node="awx-1"} 0.03401994705200195
```

**Comment by kdelee** [comment]
closing in favor of https://github.com/ansible/awx/pull/15174 ... if only because it is less surface area to test and still gets desired speedup

**Pull request closed without merging**